pyhealth/datasets/sample_dataset.py

```diff
@@ -141,45 +152,153 @@ def build(self) -> None:
                 self.output_processors[k] = self._get_processor_instance(v)
-                self.output_processors[k].fit(self.samples, k)
-        # Always process samples with the (fitted) processors
-        for sample in tqdm(self.samples, desc="Processing samples"):
-            for k, v in sample.items():
-                if k in self.input_processors:
-                    sample[k] = self.input_processors[k].process(v)
-                elif k in self.output_processors:
-                    sample[k] = self.output_processors[k].process(v)
+                self.output_processors[k].fit(iter(self.dataset), k)
         return
 
-    def __getitem__(self, index: int) -> Dict:
-        """Returns a sample by index.
+    def transform(self, sample) -> Dict:
+        """Applies the input and output processors to a sample."""
+        for k, v in sample.items():
+            if k in self.input_processors:
+                sample[k] = self.input_processors[k].process(pickle.loads(v))
+            elif k in self.output_processors:
+                sample[k] = self.output_processors[k].process(pickle.loads(v))
+            else:
+                sample[k] = pickle.loads(v)
+        return sample
 
-        Args:
-            index (int): Index of the sample to retrieve.
+    def set_shuffle(self, shuffle: bool) -> None:
+        """Sets whether to shuffle the dataset during iteration.
 
-        Returns:
-            Dict: A dict with patient_id, visit_id/record_id, and other
-            task-specific attributes as key. Conversion to index/tensor
-            will be done in the model.
+        Args:
+            shuffle (bool): Whether to shuffle the dataset.
         """
-        return self.samples[index]
+        self.dataset.set_shuffle(shuffle)
+        return
 
-    def __str__(self) -> str:
-        """Returns a string representation of the dataset.
+    def __iter__(self) -> Iterator:
+        """Returns an iterator over the dataset samples."""
+        return self.dataset.__iter__()
 
-        Returns:
-            str: A string with the dataset and task names.
-        """
+    def __getitem__(self, index: int) -> Dict:
+        """Gets a sample by index. Try to use iterator for better performance."""
+        return self.dataset.__getitem__(index)
+
+    def __str__(self) -> str:
+        """String representation of the SampleDataset."""
         return f"Sample dataset {self.dataset_name} {self.task_name}"
 
     def __len__(self) -> int:
-        """Returns the number of samples in the dataset.
+        """Returns the number of samples in the dataset."""
+        return self.dataset.__len__()
 
-        Returns:
-            int: The number of samples.
+class SampleSubset(IterableDataset):
+    """A subset of the SampleDataset.
+
+    Args:
+        sample_dataset (SampleDataset): The original SampleDataset.
+        indices (List[int]): List of indices to include in the subset.
+    """
+
```

Comment:
Add new SampleSubset to support creating subset of a SampleDataset for train/val/test split.